File: py/packages/corpora_cli/commands/workon.py
from typing import List

import typer
from corpora_client.models.corpus_file_chat_schema import CorpusFileChatSchema
from corpora_client.models.message_schema import MessageSchema
from prompt_toolkit.shortcuts import PromptSession

from corpora_cli.context import ContextObject

app = typer.Typer(help="Interactive issue creation CLI")

session = PromptSession()


@app.command()
def file(ctx: typer.Context, path: str):
    """Workon a file in the corpus."""
    c: ContextObject = ctx.obj
    c.console.print(f"Working on file: {path}", style="bold blue")
    ext = path.split(".")[-1]

    # show current file content on disk, load the content from the
    # CWD and print it with dim
    try:
        with open(path) as f:
            current_file_content = f.read() if f else ""
    except FileNotFoundError:
        current_file_content = ""
        open(path, "w").close()  # create the file if it doesn't exist

    c.console.print("Current file content:", style="bold green")
    c.console.print(current_file_content, style="dim")

    # Start with an empty message list
    messages: List[MessageSchema] = []

    if current_file_content:
        messages.append(
            MessageSchema(
                role="user",
                text=f"The original file content was:\n```{ext}\n{current_file_content}```",
            ),
        )

    # REPL loop
    while True:
        user_input = session.prompt(
            ("What to do?\n" if not messages else "How to revise?\n"),
            multiline=True,
            vi_mode=True,
        )

        if not user_input:
            c.console.print(
                "No input provided. Please try again.",
                style="yellow",
            )
            continue

        # Add the user's input as a new message
        messages.append(MessageSchema(role="user", text=user_input.strip()))

        # Send the current messages to generate a draft issue
        c.console.print("Generating revision...", style="bold blue")

        # if file doesn't exist, use empty string
        with open(".corpora/VOICE.md") as f:
            voice = f.read() if f else ""
        with open(".corpora/PURPOSE.md") as f:
            purpose = f.read() if f else ""
        with open(".corpora/STRUCTURE.md") as f:
            structure = f.read() if f else ""
        with open(f".corpora/{ext}/DIRECTIONS.md") as f:
            directions = f.read() if f else ""

        revision = c.workon_api.file(
            CorpusFileChatSchema(
                messages=messages,
                corpus_id=c.config["id"],
                path=path,
                voice=voice,
                purpose=purpose,
                structure=structure,
                directions=directions,
            ),
        )
        c.console.print(f"{revision}", style="dim")
        c.console.print(f"{path}", style="dim magenta")
        messages.append(MessageSchema(role="assistant", text=revision))

        if typer.confirm("Write file?"):
            with open(path, "w") as f:
                f.write(revision)
            c.console.print("File written!", style="green")
            continue
        c.console.print(
            "You chose not to write the file. Give more input to revise.",
            style="magenta",
        )
        continue

File: rs/core/corpora_cli/src/commands/workon.rs
use crate::context::Context;
use clap::Args;
use corpora_client::models::{CorpusFileChatSchema, MessageSchema};
use dialoguer::{theme::ColorfulTheme, Confirm};
use std::fs::{self, File};
use std::io::Write;
use std::path::Path;

#[derive(Args)]
pub struct WorkonArgs {
    #[arg(help = "Path to the file or directory")]
    pub path: String,
}

/// The `workon` command operation
pub fn run(ctx: &Context, args: WorkonArgs) {
    let path = Path::new(&args.path);
    let cwd = std::env::current_dir().unwrap();
    ctx.dim(&format!("Current working directory: {}", cwd.display()));
    let absolute_path = Path::join(&cwd, path);
    let relative_path = absolute_path
        .strip_prefix(&ctx.corpora_config.root_path)
        .unwrap();

    ctx.magenta(&format!("Working on file: {}", relative_path.display()));
    let ext = relative_path
        .extension()
        .unwrap_or_default()
        .to_str()
        .unwrap_or("");

    let current_file_content = match fs::read_to_string(path) {
        Ok(content) => content,
        Err(_) => {
            File::create(path).expect("Failed to create file");
            String::new()
        }
    };
    ctx.success("Current file content:");
    ctx.dim(&current_file_content);

    let mut messages: Vec<MessageSchema> = Vec::new();
    if !current_file_content.is_empty() {
        messages.push(MessageSchema {
            role: "user".to_string(),
            text: format!(
                "The original content of `{}` was:\n```{}\n```",
                relative_path.display(),
                current_file_content
            ),
        });
    }

    // REPL loop
    loop {
        // let user_input: String = Input::with_theme(&ColorfulTheme::default())
        //     .with_prompt(if messages.is_empty() {
        //         "What to do?"
        //     } else {
        //         "How to revise?"
        //     })
        //     .allow_empty(false)
        //     .interact_text()
        //     .unwrap();
        let user_input = ctx
            .get_user_input_via_editor(&format!(
                "Put your prompt here for {}, save and close",
                path.display()
            ))
            .expect("Failed to get user input");

        // Add the user's input as a new message
        messages.push(MessageSchema {
            role: "user".to_string(),
            text: user_input.trim().to_string(),
        });

        ctx.success("Generating revision...");

        let root_path = &ctx.corpora_config.root_path;
        let voice = fs::read_to_string(root_path.join(".corpora/VOICE.md")).unwrap_or_default();
        let purpose = fs::read_to_string(root_path.join(".corpora/PURPOSE.md")).unwrap_or_default();
        let structure =
            fs::read_to_string(root_path.join(".corpora/STRUCTURE.md")).unwrap_or_default();
        let directions =
            fs::read_to_string(root_path.join(format!(".corpora/{}/DIRECTIONS.md", ext)))
                .unwrap_or_default();

        let revision = match corpora_client::apis::workon_api::file(
            &ctx.api_config,
            CorpusFileChatSchema {
                messages: messages.clone(),
                corpus_id: ctx
                    .corpora_config
                    .id
                    .clone()
                    .expect("Failed to get corpus ID"),
                path: relative_path.to_string_lossy().to_string(),
                voice: Some(voice),
                purpose: Some(purpose),
                structure: Some(structure),
                directions: Some(directions),
            },
        ) {
            Ok(response) => response,
            Err(err) => {
                ctx.error(&format!("Failed to generate revision: {:?}", err));
                continue;
            }
        };
        ctx.dim(&revision);
        ctx.highlight(&format!("Revision for `{}`:", relative_path.display()));
        messages.push(MessageSchema {
            role: "assistant".to_string(),
            text: revision.clone(),
        });

        if Confirm::with_theme(&ColorfulTheme::default())
            .with_prompt("Write file?")
            .interact()
            .unwrap()
        {
            let mut file = File::create(path).expect("Failed to open file");
            file.write_all(revision.as_bytes())
                .expect("Failed to write file");
            ctx.success("File written!");
        } else {
            ctx.warn("You chose not to write the file. Give more input to revise.");
        }
    }
}
